File: examples/python/02-agent-basic/agent.py
"""Tier 02 — Basic agent (Python).

A minimal MCP client that acquires an Authplane-issued machine token via the
`client_credentials` grant and uses it to call a protected MCP tool on the
tier-01 server. The auth-specific code is wrapped between
`# authplane:begin` / `# authplane:end` so the `tools/loccount` tool can
audit the LOC budget for this tier.

Run via the example's Makefile (the tier-01 example must be up first):

    cp .env.example .env
    make run
    make verify
"""

# === transport / HTTP client boilerplate =====================================
import asyncio
import json
import os
import sys

import httpx


async def main() -> int:
    mcp_url = os.environ["MCP_URL"]

    # === Authplane integration ===============================================
    # authplane:begin
    from authplane import ASCredentials, AuthplaneClient

    client = await AuthplaneClient.create(
        issuer=os.environ["AUTHPLANE_ISSUER"],
        auth=ASCredentials(os.environ["CLIENT_ID"], os.environ["CLIENT_SECRET"]),
        dev_mode=True,
    )
    token = (await client.client_credentials(
        scopes=["mcp:echo"], resources=[os.environ["RESOURCE_URI"]],
    )).access_token
    # authplane:end

    # === call the MCP tool ===================================================
    # JSON-RPC `initialize` over the streamable-http transport. The full
    # `tools/call` interaction requires a stateful session handshake; the
    # initialize step alone is enough to prove that the bearer token cleared
    # the resource server's auth chain — without the token the same request
    # returns 401.
    try:
        async with httpx.AsyncClient(timeout=10.0) as http:
            resp = await http.post(
                mcp_url,
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json, text/event-stream",
                },
                content=json.dumps({
                    "jsonrpc": "2.0", "id": 1, "method": "initialize",
                    "params": {
                        "protocolVersion": "2024-11-05", "capabilities": {},
                        "clientInfo": {"name": "tier02-agent", "version": "1.0"},
                    },
                }),
            )
        if resp.status_code not in (200, 202):
            print(f"MCP call failed: HTTP {resp.status_code}\n{resp.text}", file=sys.stderr)
            return 1
        # The streamable-http transport frames replies as either raw JSON or
        # as an SSE `data:` event. Either way the initialize response carries
        # the server's name in `serverInfo.name`.
        body = resp.text
        print(body)
        if "demo-server" not in body:
            print("MCP response did not contain expected serverInfo.name", file=sys.stderr)
            return 1
        print("authenticated MCP initialize OK", flush=True)
        return 0
    finally:
        await client.aclose()


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: examples/typescript/02-agent-basic/agent.ts
/**
 * Tier 02 — Basic agent (TypeScript).
 *
 * A minimal MCP client that acquires an access token from Authplane via the
 * `client_credentials` grant and uses it to call the tier-01 MCP server's
 * `echo` tool. Auth-specific lines are wrapped between
 * `// authplane:begin` / `// authplane:end` so `tools/loccount` can audit
 * the LOC budget for this tier.
 *
 * Run via the example's Makefile (tier 01 must already be running):
 *
 *     # In examples/typescript/01-mcp-server-basic
 *     make run
 *
 *     # In examples/typescript/02-agent-basic
 *     cp .env.example .env
 *     make verify
 */

// === transport / HTTP client boilerplate =====================================
const MCP_URL = process.env.MCP_URL ?? "http://mcp-server:8080/mcp";
const ECHO_TEXT = process.env.ECHO_TEXT ?? "hello from tier 02";

async function mcpCall(accessToken: string, method: string, params: unknown, id: number): Promise<unknown> {
  const res = await fetch(MCP_URL, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${accessToken}`,
      "Content-Type": "application/json",
      "Accept": "application/json, text/event-stream",
    },
    body: JSON.stringify({ jsonrpc: "2.0", id, method, params }),
  });
  if (!res.ok) {
    throw new Error(`MCP ${method} returned HTTP ${res.status}: ${await res.text()}`);
  }
  // The streamable-http transport may reply as SSE; pull the first `data:` line.
  const body = await res.text();
  const dataLine = body.split("\n").find((l) => l.startsWith("data: "));
  const payload = dataLine ? dataLine.slice("data: ".length) : body;
  return JSON.parse(payload);
}

// === Authplane integration ===================================================
// authplane:begin
import { AuthplaneClient } from "@authplane/sdk/core";
const ap = await AuthplaneClient.create({
  issuer: process.env.AUTHPLANE_ISSUER!,
  auth: { clientId: process.env.AUTHPLANE_CLIENT_ID!, clientSecret: process.env.AUTHPLANE_CLIENT_SECRET! },
});
const token = await ap.clientCredentials(["mcp:echo"], [process.env.AUTHPLANE_RESOURCE!]);
// authplane:end

// === call the MCP tool =======================================================
const initResult = await mcpCall(token.accessToken, "initialize", {
  protocolVersion: "2024-11-05",
  capabilities: {},
  clientInfo: { name: "tier-02-agent", version: "1.0" },
}, 1);
console.log(`[agent] initialize OK: ${JSON.stringify(initResult)}`);

const echoResult = await mcpCall(token.accessToken, "tools/call", {
  name: "echo",
  arguments: { text: ECHO_TEXT },
}, 2);
console.log(`[agent] echo OK: ${JSON.stringify(echoResult)}`);

await ap.close();
